optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java

```diff
@@ -15,12 +18,16 @@ public class PythonSolver {
      */
     public static boolean onlyUseJavaSetters = false;
```

Comment:
How public is this API?

I'd prefer it if we didn't use any mutable static structures (shared state) for communication between a user and the program. In this case, I suggest we could put this boolean as a method argument where it is necessary.

Reply:
Not public, the user doesn't know it exist. In case you wondering what all the `# noqa` comments are for, it so the IDE typechecker does not raise warnings/errors that it cannot find these packages/classes. That being said, nothing can stop a determined Python programmer from touching internals, but they wouldn't know these internals exist without looking at the source code.

Reply:
Got it. Still, I'd rather we not use shared state. See if you can get rid of it, please.

Reply:
Very hard to remove this shared state, since it is how the `ConstraintProvider` communicates to the `Solver` if it has been successfully FULLY translated to `Java` (and thus can completely avoid going into Python). It is used here: https://github.com/optapy/optapy/blob/db84e20197616175fd606ad371bb63f9caf26565/optapy-core/src/main/python/annotations.py#L664-L678 and here: https://github.com/optapy/optapy/blob/db84e20197616175fd606ad371bb63f9caf26565/optapy-core/src/main/python/optaplanner_api_wrappers.py#L37-L71 ; in short, we cannot know the value until `defineConstraints` is called (since that what compiles all the Python functions), and `definedConstraints` is called during `SolverFactory`/`SolverManager` creation. `PythonSolver` needs a way to read the `ConstraintProvider` instance itself in order to remove that shared state (since `solve` in Python also does not have access to the `ConstraintProvider` that is being used. This field is used to determine what setter to use in ALL domain objects; it will either be a no-op (`(ref, field, value) -> {}`) or it will call CPython (`(ref, field, value) -> CPython.setAttributeOnObject(ref, field, value)`). The objects are wrapped at solve time, and depending on the score function, may need to go to CPython, or can stay in pure java.